### PR TITLE
Fix mod added Villager Trades failing serialization

### LINKED ISSUE #1281: [1.21] Using BasicItemListing to add villager trades throw an error on game save
**Minecraft Version:** 1.21

**NeoForge Version:** 21.0.76

**Logs:**  https://gist.github.com/OrderedChaos-Dev/fef69c74be38cfdee41d089ed20c2c88

**Steps to Reproduce:**
 1. Add a trade via `WandererTradesEvent` / `VillagerTradesEvent` with `net.neoforged.neoforge.common.BasicItemListing`. Use any of the constructors in `BasicItemListing` that doesn't have `price2`
 2. Load game and find a villager with that trade
 3. Save and check logs

**Description of issue:**  
When using `net.neoforged.neoforge.common.BasicItemListing` to add villager trades, the game will throw an error when trying to save villagers with those trades. Saving continues as normal but the villagers will be deleted.  

The error thrown is: `Caused by: java.lang.IllegalStateException: Value must be positive: 0; Item must not be minecraft:air`  

I suspect this is because `BasicItemListing.getOffer()` creates a `MerchantOffer` with `Optional.of(cost2)`, with `cost2` being `ItemStack.EMPTY` instead of `Optional.empty()` for trades with no 2nd cost item, which fails validation when trying to serialize the trade - see `net.minecraft.world.item.trading.ItemCost`.`CODEC`:  

<pre><code>// "Item must not be minecraft:air"
ItemStack.ITEM_NON_AIR_CODEC.fieldOf("id").forGetter(ItemCost::item),
// "Value must be positive: 0"
ExtraCodecs.POSITIVE_INT.fieldOf("count").orElse(1).forGetter(ItemCost::count),
</code></pre>

### PR DESCRIPTION
**Pull request by TelepathicGrunt**
Closes https://github.com/neoforged/NeoForge/issues/1281

Tested by writing a GameTest that spawns a Villager, adds a modded trade to it, then I spawned it in game and paused the game. There was a save serialization error as a results.

By setting the MerchantOffer's second item cost to Optional.EMPTY if it was given no second trade, repeating the same steps produces no error anymore.

Bonus, the GameTest itself will make sure our Villager Trade event continues to inject new trades.

### CODE REVIEW

File: tests/src/main/java/net/neoforged/neoforge/debug/entity/TradeTests.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+import net.neoforged.testframework.annotation.ForEachTest;
+import net.neoforged.testframework.annotation.TestHolder;
+import net.neoforged.testframework.gametest.EmptyTemplate;
+
+@ForEachTest(groups = TradeTests.GROUP)
+public class TradeTests {
+    public static final String GROUP = "level.entity.trades";
+    private static final String MOD_ID = "villager_trade_test";
+
+    @Mod(value = MOD_ID)
```
Comment (Matyrobbrt) [inline tests/src/main/java/net/neoforged/neoforge/debug/entity/TradeTests.java]: Why is this a mod and not a static event listener?

**Review by Matyrobbrt — CHANGES_REQUESTED** [review]

File: tests/src/main/java/net/neoforged/neoforge/debug/entity/TradeTests.java (hunk 1 above)
Comment (TelepathicGrunt) [reply tests/src/main/java/net/neoforged/neoforge/debug/entity/TradeTests.java]: Don't all the subscribe annotations require it to be attached to a mod. I cannot find an example of a static event listener in any of the new tests so I will need an example of what you mean if not the annotations

Comment (TelepathicGrunt) [reply tests/src/main/java/net/neoforged/neoforge/debug/entity/TradeTests.java]: Test fixed and made more robust to ensure we don't lost vanilla trades by mistake

File: tests/src/main/java/net/neoforged/neoforge/debug/entity/TradeTests.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+        test.onGameTest(helper -> helper.startSequence(() -> helper.spawn(EntityType.VILLAGER, new BlockPos(1, 1, 1)))
+                .thenExecute(villager -> {
+                    // Set villager to level 6 trades
+                    villager.setVillagerData(new VillagerData(VillagerType.PLAINS, VillagerProfession.WEAPONSMITH, 0));
+                    ((VillagerAccessor) villager).neoforge$callIncreaseMerchantCareer();
+                    ((VillagerAccessor) villager).neoforge$callIncreaseMerchantCareer();
+                    ((VillagerAccessor) villager).neoforge$callIncreaseMerchantCareer();
+                    ((VillagerAccessor) villager).neoforge$callIncreaseMerchantCareer();
+                    ((VillagerAccessor) villager).neoforge$callIncreaseMerchantCareer();
+                    ((VillagerAccessor) villager).neoforge$callIncreaseMerchantCareer();
```
Comment (Matyrobbrt) [inline tests/src/main/java/net/neoforged/neoforge/debug/entity/TradeTests.java]: Should be reflection instead.

**Review by Matyrobbrt — CHANGES_REQUESTED** [review]

File: tests/src/main/java/net/neoforged/neoforge/debug/entity/TradeTests.java (hunk 2 above)
Comment (TelepathicGrunt) [reply tests/src/main/java/net/neoforged/neoforge/debug/entity/TradeTests.java]: Get reflected on! *dabs*

**Review by shartte — DISMISSED** [review]

**Review by Matyrobbrt — DISMISSED** [review]

**Review by shartte — APPROVED** [review]

**Pull request merged**